atlasdb-client/src/main/java/com/palantir/atlasdb/keyvalue/impl/InMemoryKeyValueService.java

```diff
@@ -362,31 +371,103 @@ public void putWithTimestamps(TableReference tableRef, Multimap<Cell, Value> val
-            throws KeyAlreadyExistsException {
-        putInternal(tableRef, KeyValueServices.toConstantTimestampValues(values.entrySet(), 0), true);
+    public void checkAndSet(CheckAndSetRequest request) throws CheckAndSetException {
+        Table table = getTableMap(request.table());
+        Cell cell = request.cell();
+        Optional<byte[]> oldValue = request.oldValue();
+        byte[] contents = request.newValue();
+
+        Key key = getKey(table, cell, AtlasDbConstants.TRANSACTION_TS);
+        if (oldValue.isPresent()) {
+            byte[] storedValue = table.entries.get(key);
+            boolean succeeded = Arrays.equals(storedValue, oldValue.get())
+                    && table.entries.replace(key, storedValue, copyOf(contents));
+            if (!succeeded) {
+                byte[] actual = table.entries.get(key); // Refetch, something may have happened between get and replace
+                throwCheckAndSetException(cell, oldValue.get(), actual);
+            }
+        } else {
+            byte[] oldContents = putIfAbsent(table, key, contents);
+            if (oldContents != null) {
+                String actual = new String(oldContents, StandardCharsets.UTF_8);
+                String msg = String.format("Unexpected value for this key. Wanted (empty), got %s", actual);
+                throw new CheckAndSetException(msg, cell, null, ImmutableList.of(oldContents));
+            }
+        }
+    }
+
+    private void throwCheckAndSetException(Cell cell, byte[] expected, byte[] actual) {
+        String expectedStr = new String(expected, StandardCharsets.UTF_8);
+        if (actual != null) {
+            String actualStr = new String(actual, StandardCharsets.UTF_8);
+            String msg = String.format("Unexpected value for this key. Wanted %s, got %s", expectedStr, actualStr);
+            throw new CheckAndSetException(msg, cell, expected, ImmutableList.of(actual));
+        } else {
+            String msg = String.format("Unexpected value for this key. Wanted %s, got (empty)", expectedStr);
+            throw new CheckAndSetException(msg, cell, expected, ImmutableList.of());
+        }
+    }
+
+    private List<byte[]> getStoredValues(Cell cell, TableReference tableRef) {
+        Map<Cell, Value> storedValue = get(tableRef, ImmutableMap.of(cell, AtlasDbConstants.TRANSACTION_TS + 1));
+        return storedValue.values().stream().map(Value::getContents).collect(Collectors.toList());
+    }
+
+    private boolean valuesMatch(Map<Cell, Value> storedValue, Optional<byte[]> expectedValue) {
+        if (!expectedValue.isPresent()) {
+            return storedValue.isEmpty();
+        }
+
+        return storedValue.entrySet().size() == 1
+                && Arrays.equals(expectedValue.get(), Iterables.getOnlyElement(storedValue.values()).getContents());
+    }
+
+    private String getStringRepresentation(Optional<byte[]> oldValue) {
```

Comment:
looks like this might have been refactored away? or part of the work on generating the exception method, but isn't used at the moment.